### ISSUE TITLE
403 Forbidden : Access is denied

### ISSUE THREAD

**Issue by rollingpanda007** [usnistgov/ACVP#968]
I have the right certificate and key for establishing connection with the acvp server,
but on running the following command : acvp_app -aes the following error is obtained:

 ACV_SERVER:     demo.acvts.nist.gov
    ACV_PORT:       443
    ACV_URI_PREFIX: /acvp/v1/
    ACV_CA_FILE:    libacvp\certs\acvp.nist.gov.crt
    ACV_CERT_FILE:  Demo.cer
    ACV_KEY_FILE:   demo.key

***ACVP [STATUS][acvp_login:2509]--> Logging in...
***ACVP [ERR][log_network_status:1491]--> 403 error received from server. Message:
***ACVP [ERR][log_network_status:1492]--> <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
<title>403 - Forbidden: Access is denied.</title>
<style type="text/css">

</style>
</head>
<body>
<div id="header"><h1>Server Error</h1></div>
<div id="content">
 <div class="content-container"><fieldset>
  <h2>403 - Forbidden: Access is denied.</h2>
  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>
 </fieldset></div>
</div>
</body>
</html>

***ACVP [STATUS][acvp_login:2521]--> Login Send Failed
***ACVP [ERR][acvp_run:3200]--> Failed to login with ACVP server

Also on running this command : curl -v --cacert acvp.nist.gov.crt --key yourkey.key --cert your_certificate.cer -H "Content-Type: application/json" -X POST https://demo.acvts.nist.gov/acvp/v1/login

Maybe the output of this command helps.  
OST https://demo.acvts.nist.gov/acvp/v1/login                                            
* Rebuilt URL to: file/                                                                  
* Could not resolve host: file                                                           
* Closing connection 0                                                                   
curl: (6) Could not resolve host: file                                                   
* Rebuilt URL to: for/                                                                   
* Could not resolve host: for                                                            
* Closing connection 1                                                                   
curl: (6) Could not resolve host: for                                                    
* Rebuilt URL to: access/                                                                
* Could not resolve host: access                                                         
* Closing connection 2                                                                   
curl: (6) Could not resolve host: access                                                 
* Rebuilt URL to: to/                                                                    
* Could not resolve host: to                                                             
* Closing connection 3                                                                   
curl: (6) Could not resolve host: to                                                     
* Rebuilt URL to: ACVTS/                                                                 
* Could not resolve host: ACVTS                                                          
* Closing connection 4                                                                   
curl: (6) Could not resolve host: ACVTS                                                  
* Rebuilt URL to: demo20200813152440\Demo.cer/               
* Could not resolve host: demo20200813152440\Demo.cer        
* Closing connection 5                                                                   
curl: (6) Could not resolve host: demo20200813152440\IntrinsicID_Ashwini_Pohekar_            
* TCP_NODELAY set                                                                        
* Connected to demo.acvts.nist.gov (52.204.191.193) port 443 (#6)                             
* schannel: SSL/TLS connection with demo.acvts.nist.gov port 443 (step 1/3)                   
* schannel: checking server certificate revocation                                            
* schannel: sending initial handshake data: sending 184 bytes...                              
* schannel: sent initial handshake data: sent 184 bytes                                       
* schannel: SSL/TLS connection with demo.acvts.nist.gov port 443 (step 2/3)                   
* schannel: failed to receive handshake, need more data                                       
* schannel: SSL/TLS connection with demo.acvts.nist.gov port 443 (step 2/3)                   
* schannel: encrypted data got 3731                                                           
* schannel: encrypted data buffer: offset 3731 length 4096                                    
* schannel: sending next handshake data: sending 182 bytes...                                 
* schannel: SSL/TLS connection with demo.acvts.nist.gov port 443 (step 2/3)                   
* schannel: encrypted data got 107                                                            
* schannel: encrypted data buffer: offset 107 length 4096                                     
* schannel: SSL/TLS handshake complete                                                        
* schannel: SSL/TLS connection with demo.acvts.nist.gov port 443 (step 3/3)                   
* schannel: stored credential handle in session cache                                         
> POST /acvp/v1/login HTTP/1.1                                                                
> Host: demo.acvts.nist.gov                                                                   
> User-Agent: curl/7.55.1-DEV                                                                 
> Accept: */*                                                                                 
> Content-Type: application/json                                                              
>                                                                                             
* schannel: client wants to read 102400 bytes                                                 
* schannel: encdata_buffer resized 103424                                                     
* schannel: encrypted data buffer: offset 0 length 103424                                     
* schannel: encrypted data got 597                                                            
* schannel: encrypted data buffer: offset 597 length 103424                                   
* schannel: decrypted data length: 527                                                        
* schannel: decrypted data added: 527                                                         
* schannel: decrypted data cached: offset 527 length 102400                                   
* schannel: encrypted data buffer: offset 0 length 103424                                     
* schannel: decrypted data buffer: offset 527 length 102400                                   
* schannel: schannel_recv cleanup                                                             
* schannel: decrypted data returned 527                                                       
* schannel: decrypted data buffer: offset 0 length 102400                                     
< HTTP/1.1 411 Length Required                                                                
< Content-Type: text/html; charset=us-ascii                                                   
< Server: Microsoft-HTTPAPI/2.0                                                               
< Date: Mon, 24 Aug 2020 13:24:41 GMT                                                         
< Connection: close                                                                           
< Content-Length: 344                                                                         
<                                                                                             
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">     
<HTML><HEAD><TITLE>Length Required</TITLE>                                                    
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>                 

Is there some way this issue can be resolved. Further information : The connection is not being established through proxy server.

**Comment by celic** [comment]
What is the email on the certificate subject?

**Comment by jarnold01** [comment]
Hi @rollingpanda007 

First, I would suggest changing the ACV_CA_FILE variable to point to the mozilla_trust_anchors.pem file.
Next, please confirm that you are syncing to a time server via ntp.
Next, please confirm that you are referencing your TOTP see value (please do not post the actual value in your reply) in the ACV_TOTP_SEED variable (note, the TOTP value itself, NOT the path to the TOTP.txt file)
Finally, on your next attempt, please type the 'date' command so I have a time/date stamp to be able to look in the server logs to find your connection attempt.

Thanks.

**Comment by jarnold01** [comment]
Also, please try using this command (which does not require OTP) to confirm that communication is working properly:

./app/acvp_app --get /acvp/v1/algorithms

You should receive back the list of supported algorithms in the response.

**Comment by rollingpanda007** [comment]
Hi @jarnold 

On running acvp_app --get /acvp/v1/algorithms, I get the following output 
λ acvp_app --aes

Using the following parameters:

    ACV_SERVER:     demo.acvts.nist.gov
    ACV_PORT:       443
    ACV_URI_PREFIX: /acvp/v1/
    ACV_CA_FILE:    C:\Users\pohekar\Desktop\demo-1-1-0\libacvp-libacvp_1_1_0-throttle\certs\mozzila_trust_anchors.pem
    ACV_CERT_FILE:  C:\Users\pohekar\Downloads\Re_CSR file for access to ACVTS demo20200813152440\IntrinsicID_Ashwini_Pohekar_Demo.cer
    ACV_KEY_FILE:   C:\Users\pohekar\Documents\libacvp_cert\Ashwini.Pohekar_public.key

***ACVP [STATUS][acvp_login:2509]--> Logging in...
***ACVP [ERR][log_network_status:1491]--> 403 error received from server. Message:
***ACVP [ERR][log_network_status:1492]--> <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
<title>403 - Forbidden: Access is denied.</title>
<style type="text/css">

</style>
</head>
<body>
<div id="header"><h1>Server Error</h1></div>
<div id="content">
 <div class="content-container"><fieldset>
  <h2>403 - Forbidden: Access is denied.</h2>
  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>
 </fieldset></div>
</div>
</body>
</html>

***ACVP [STATUS][acvp_login:2521]--> Login Send Failed
***ACVP [ERR][acvp_run:3200]--> Failed to login with ACVP server

**Comment by rollingpanda007** [comment]
Hi @jarnold01 

Here is another attempt to connect with the server on running the following command : acvp_app.exe --aes --verbose
Time stamp : 16:20:39.53 CET
Command output : Using the following parameters:                                                                                                                                                                  
                                                                                                                                                                                                 
    ACV_SERVER:     demo.acvts.nist.gov                                                                                                                                                          
    ACV_PORT:       443                                                                                                                                                                          
    ACV_URI_PREFIX: /acvp/v1/                                                                                                                                                                    
    ACV_CA_FILE:    libacvp-libacvp_1_1_0-throttle\certs\mozzila_trust_anchors.pem                                                                           
    ACV_CERT_FILE:  Demo.cer                                                           
    ACV_KEY_FILE:   demo.key                                                                                                           
                                                                                                                                                                                                 
***ACVP [STATUS][acvp_login:2509]--> Logging in...                                                                                                                                               
***ACVP [INFO][acvp_http_user_agent_handler:483]--> HTTP User-Agent: libacvp/1.0;Windows 10 Pro;18362.19h1_release.190318-1202;x86_64;Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz;MSVC/191627043    
                                                                                                                                                                                                 
***ACVP [INFO][log_network_status:1464]--> POST Login...                                                                                                                                         
        Status: 403                                                                                                                                                                              
        Url: https://demo.acvts.nist.gov:443/acvp/v1/login                                                                                                                                       
        Resp: Recieved                                                                                                                                                                           
                                                                                                                                                                                                 
***ACVP [ERR][log_network_status:1491]--> 403 error received from server. Message:                                                                                                               
***ACVP [ERR][log_network_status:1492]--> <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">                                          
<html xmlns="http://www.w3.org/1999/xhtml">                                                                                                                                                      
<head>                                                                                                                                                                                           
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>                                                                                                                        
<title>403 - Forbidden: Access is denied.</title>                                                                                                                                                
<style type="text/css">                                                                                                                                                                          
<!--                                                                                                                                                                                             
body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}                                                                                              
fieldset{padding:0 15px 10px 15px;}                                                                                                                                                              
h1{font-size:2.4em;margin:0;color:#FFF;}                                                                                                                                                         
h2{font-size:1.7em;margin:0;color:#CC0000;}                                                                                                                                                      
h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;}                                                                                                                                             
#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:"trebuchet MS", Verdana, sans-serif;color:#FFF;                                                                               
background-color:#555555;}                                                                                                                                                                       
#content{margin:0 0 0 2%;position:relative;}                                                                                                                                                     
.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}                                                                                                     
-->                                                                                                                                                                                              
</style>                                                                                                                                                                                         
</head>                                                                                                                                                                                          
<body>                                                                                                                                                                                           
<div id="header"><h1>Server Error</h1></div>                                                                                                                                                     
<div id="content">                                                                                                                                                                               
 <div class="content-container"><fieldset>                                                                                                                                                       
  <h2>403 - Forbidden: Access is denied.</h2>                                                                                                                                                    
  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>                                                                                    
 </fieldset></div>                                                                                                                                                                               
</div>                                                                                                                                                                                           
</body>                                                                                                                                                                                          
</html>                                                                                                                                                                                          
                                                                                                                                                                                                 
***ACVP [STATUS][acvp_login:2521]--> Login Send Failed                                                                                                                                           
***ACVP [ERR][acvp_run:3200]--> Failed to login with ACVP server

**Comment by jarnold01** [comment]
Thanks for providing the timestamp information. I searched through the web server logs and saw your attempted connections:

2020-08-25 10:21:37 xxx.xxx.xxx.xxx POST /acvp/v1/login - 443 - xxx.xxx.xxx.xxx libacvp/1.0;Windows+10+Pro;18362.19h1_release.190318-1202;x86_64;Intel(R)+Core(TM)+i7-7700HQ+CPU+@+2.80GHz;MSVC/192729111 - 403 7 5 1525 304 203
2020-08-25 10:21:58 xxx.xxx.xxx.xxx POST /acvp/v1/login - 443 - xxx.xxx.xxx.xxx libacvp/1.0;Windows+10+Pro;18362.19h1_release.190318-1202;x86_64;Intel(R)+Core(TM)+i7-7700HQ+CPU+@+2.80GHz;MSVC/192729111 - 403 7 5 1525 304 187
2020-08-25 10:23:05 xxx.xxx.xxx.xxx POST /acvp/v1/login - 443 - xxx.xxx.xxx.xxx libacvp/1.0;Windows+10+Pro;18362.19h1_release.190318-1202;x86_64;Intel(R)+Core(TM)+i7-7700HQ+CPU+@+2.80GHz;MSVC/192729111 - 403 7 5 1525 304 203

The server logs are indicating that the supplied certificate is not valid (403.7). Is there an outbound corporate proxy that you are forced to communicate through, by any chance? If so, it may be replacing your certificate with it's own certificate and presenting that to the server instead, which would fail (we just recently determined this was the issue for another user.) 

Is the PEM formatting of your certificate file intact? (if you ever opened/edited your cert file on Windows, it's possible that ^M (Windows CR/LF) characters were added to each line in the file, which could be causing a validation issue.)

For now just continue testing with the **./app/acvp_app --get /acvp/v1/algorithms** command, please, until we figure out the certificate issue. After that is resolved, we can move onto the next step.

**Comment by rollingpanda007** [comment]
Hi @jarnold01  

I am not communicating from behind a proxy server. Also, I am quite sure I did not open the certificate files. Is there any way to verify if they are still intact/valid ? 
I will keep verifying with the acvp_app --get /acvp/v1/algorithms command for now. 

Thanks!

**Comment by jarnold01** [comment]
You are welcome to zip up your demo.cer file and then send the zip file to me via https://nfiles.nist.gov if you'd like. I can verify it for you.

**Comment by rollingpanda007** [comment]
Hi @jarnold01 

I have sent the zip file via nfiles. Please let me know if the certificate still works.

**Comment by jarnold01** [comment]
Hi @rollingpanda007 - Thanks for sending it over. The certificate validates just fine.

**Comment by rollingpanda007** [comment]
Hi @jarnold01 

This rules out both possibilities then. The certificate is valid and also while establishing communication with the server there is no proxy server in between. Then what could be causing the 403 forbidden access is denied error.

**Comment by jarnold01** [comment]
Hi @rollingpanda007 

It is definitely possibly an issue with the build process on your system, sure. I've only ever build against openssl 1.1.1 FIPS, for example, and only within a Linux docker container at that. I've sent you a reply via nfiles a short while ago with a suggested next step to determine if it's a client-side issue or not, so have a look at that and let me know.

**Comment by rollingpanda007** [comment]
Hi @jarnold01 

On the openssl page here : https://www.openssl.org/source/,
The openssl 3.0.0 which is a pre-release version under development is the one which supports the FIPS compliant build for openssl 1.1.1. Is this the version that you used for building your libacvp client.

Other than that, only the openssl 1.0.2 FIPS complaint version which is available. 
Also, could it be because of the version of libcurl being used ? If so, can you also mention the version of libcurl that you are using ?

**Comment by jarnold01** [comment]
Hi @rollingpanda007 

I looked back at my notes for clarity, and I effectively followed the steps [here](https://github.com/cisco/libacvp/tree/master/docker) to build libacvp with openssl1.1.1 into a Linux docker container.

As this is different from your platform (apparently on Windows), which I haven't done, I can't really speak to it. I did build the above Linux docker container to run on Windows 10 using Docker Desktop for Windows, but I haven't done a native build.

Perhaps someone else following this issue who has built libacvp directly in Windows would care to comment?
